Update CodeParrot readme to include training in Megatron (#17798)
huggingface/transformers: examples/research_projects/codeparrot/README.md

The numbers were obtained by sampling with `T = [0.2, 0.6, 0.8]` and picking the best value for each metric. Both CodeParrot 🦜 models are still underfitted and longer training would likely improve the performance.

## Demo
Give the model a shot yourself! There are three demos to interact with CodeParrot 🦜:
- [Code generation](https://huggingface.co/spaces/codeparrot/codeparrot-generation)
- [Code highlighting](https://huggingface.co/spaces/codeparrot/codeparrot-highlighting)
- [Comparison to other code models](https://huggingface.co/spaces/codeparrot/loubnabnl/code-generation-models)

## Training with Megatron
[Megatron](https://github.com/NVIDIA/Megatron-LM) is a framework developed by NVIDIA for training large transformer models. We found that the training of CodeParrot is faster there. Below we explain how to use it.

> 

Reviewer: Maybe we can talk a bit about the trad-offs here: while the codeparrot code is easy to follow and modify to your needs the megatron framework lets you train models faster.

### Setup

> You can pull an NVIDIA PyTorch Container that comes with all the required installations, they are available on NGC https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch with their [documentation](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/index.html):

Reviewer: ```suggestion
You can pull an NVIDIA PyTorch Container that comes with all the required installations from [NGC](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch). See [documentation](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/index.html) for more details:
```

> You can pull an NVIDIA PyTorch Container that comes with all the required installations from [NGC](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch). See [documentation](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/index.html) for more details:

Reviewer 2: We don't run things on a container ... but if you tried it and it works then it's probably the easiest thing to get running.

Author: Indeed the container works and avoids having to install everything and handle the dependencies

> After installing Docker, you can use the command below to run the container where `xx.xx` denotes your Docker version, and clone [Megatron repository](https://github.com/NVIDIA/Megatron-LM) inside it:

Reviewer: ```suggestion
With the following Docker command you can run the container (`xx.xx` denotes your Docker version), and clone [Megatron repository](https://github.com/NVIDIA/Megatron-LM) into it:
```

```bash
docker run --gpus all -it --rm nvcr.io/nvidia/pytorch:xx.xx-py3
git clone https://github.com/NVIDIA/Megatron-LM
```

> You also need to copy to the container the vocabulary file and merges table of the tokenizer that you trained on code, you can also find these files in [vocab.json](https://huggingface.co/codeparrot/codeparrot/raw/main/vocab.json) and [merges.txt](https://huggingface.co/codeparrot/codeparrot/raw/main/merges.txt).

Reviewer: ```suggestion
You also need to add the vocabulary file and merges table of the tokenizer that you trained on code into the container. You can also find these files in [vocab.json](https://huggingface.co/codeparrot/codeparrot/raw/main/vocab.json) and [merges.txt](https://huggingface.co/codeparrot/codeparrot/raw/main/merges.txt).
```

```bash
sudo docker cp vocab.json CONTAINER_ID:/workspace/Megatron-LM
sudo docker cp merges.txt CONTAINER_ID:/workspace/Megatron-LM

> ```

Author: CodParrot uses GPT2 tokenizer but trained on code, I’m not sure I understand your comment since we need to specify the path for vocab.json and merges.txt to run the preprocessing and training

> ```

Reviewer 2: Ah I was confused, in Meg-DS we added support for HF `tokenizers` and so we can use any tokenizer from the HF system

DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE --nnodes $NNODES --node_rank $NODE_RANK --master_addr $MASTER_ADDR --master_port $MASTER_PORT"
CHECKPOINT_PATH=/workspace/Megatron-LM/experiments/codeparrot-small
VOCAB_FILE=vocab.json
MERGE_FILE=merges.txt
DATA_PATH=codeparrot_content_document
GPT_ARGS="--num-layers 12
--hidden-size 768
--num-attention-heads 12
--seq-length 1024
--max-position-embeddings 1024
--micro-batch-size 12
--global-batch-size 192
--lr 0.0005
--train-iters 150000
--lr-decay-iters 150000
--lr-decay-style cosine
--lr-warmup-iters 2000
--weight-decay .1
--adam-beta2 .999
--fp16
--log-interval 10
--save-interval 2000
--eval-interval 200
--eval-iters 10
"
TENSORBOARD_ARGS="--tensorboard-dir experiments/tensorboard"
python3 -m torch.distributed.launch $DISTRIBUTED_ARGS \
        pretrain_gpt.py \
        --tensor-model-parallel-size 1 \
        --pipeline-model-parallel-size 1 \
        $GPT_ARGS \
        --vocab-file $VOCAB_FILE \
        --merge-file $MERGE_FILE \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH \
        --data-path $DATA_PATH \
        $TENSORBOARD_ARGS
```
The training takes almost 12 hours in this setting.

> ### Evaluation

Reviewer: I would probably call this conversion:
```suggestion
### Convert model to `transformers`
```

> ### Convert model to `transformers`

Author: I actually tried the conversion and it seems to work for our case because we didn't change anything in the architecture of GPT2 and the conversion script I'm using here was designed for it

> ### Convert model to `transformers`

Reviewer 2: Yeah I understand that there exists a combination that works, doesn't mean that all configuration are. But okay, we do probably support quite a lot of config so it should be fine

> ### Convert model to `transformers`

Author: Ah I see, I thought it was because you were using a different architecture in BS and not  because of the config combination, but as you said it seems to work for this case at least for gpt2-small and gpt2-large, didn't try the 1.5B one

> To evaluate the model on HumanEval, you can convert it to `transformers` following this tutorial: https://huggingface.co/nvidia/megatron-gpt2-345m. For instance, after the training is finished you can copy the weights of the last iteration 150k and convert the `model_optim_rng.pt` file to a `pytorch_model.bin` file that is supported by `transformers`.

Reviewer: ```suggestion
After training we want to use the model in `transformers` e.g. to evaluate it on HumanEval. You can convert it to `transformers` following [this](https://huggingface.co/nvidia/megatron-gpt2-345m) tutorial. For instance, after the training is finished you can copy the weights of the last iteration 150k and convert the `model_optim_rng.pt` file to a `pytorch_model.bin` file that is supported by `transformers`.
```

```bash
mkdir -p nvidia/megatron-codeparrot-small
sudo docker cp CONTAINER_ID:/workspace/Megatron-LM/experiments/codeparrot-small/iter_0150000/mp_rank_00/model_optim_rng.pt nvidia/megatron-codeparrot-small
git clone https://github.com/huggingface/transformers.git
git clone https://github.com/NVIDIA/Megatron-LM.git
export PYTHONPATH=Megatron-LM

> python transformers/src/transformers/models/megatron_gpt2/convert_megatron_gpt2_checkpoint.py nvidia/megatron-gpt2-345m/model_optim_rng.pt

Reviewer: That looks like another model (345m?). Is this a typo?

Author: Yes it's a typo!
